source/contracts/reporting/Auction.sol

```diff
@@ -0,0 +1,239 @@
+        bootstrapped = true;
+
+        uint256 _auctionRepBalanceTarget = reputationToken.totalSupply() / Reporting.getAuctionTargetSupplyDivisor();
+        uint256 _repBalance = reputationToken.balanceOf(this);
+
+        if (_repBalance < _auctionRepBalanceTarget) {
+            reputationToken.mintForAuction(_auctionRepBalanceTarget.sub(_repBalance));
+        } else {
+            reputationToken.burnForAuction(_repBalance.sub(_auctionRepBalanceTarget));
+        }
+
+        initialAttoRepBalance = _auctionRepBalanceTarget;
+        initialAttoEthBalance = address(this).balance;
+
+        currentAttoRepBalance = initialAttoRepBalance;
+        currentAttoEthBalance = initialAttoEthBalance;
+
+        currentUpperBoundRepPriceInAttoEth = 0;
+        currentLowerBoundRepPriceInAttoEth = 0;
+
+        currentAuctionIndex = _currentAuctionIndex;
+
+        initialRepSalePrice = lastRepPrice.mul(Reporting.getAuctionInitialPriceMultiplier());
+        initialEthSalePrice = Reporting.getAuctionInitialPriceMultiplier().mul(10**36).div(lastRepPrice);
+        return true;
+    }
+
+    function tradeRepForEth(uint256 _attoEthAmount) public initializeNewAuctionIfNeeded returns (bool) {
+        require(!bootstrapMode);
+        require(currentAttoEthBalance > 0);
+        require(_attoEthAmount > 0);
+        _attoEthAmount = _attoEthAmount.min(currentAttoEthBalance);
+        uint256 _ethPriceInAttoRep = getEthSalePriceInAttoRep();
+        uint256 _attoRepCost = _attoEthAmount.mul(_ethPriceInAttoRep) / 10**18;
+        reputationToken.trustedAuctionTransfer(msg.sender, this, _attoRepCost);
+        msg.sender.transfer(_attoEthAmount);
+        uint256 _attoEthSold = initialAttoEthBalance.sub(currentAttoEthBalance);
+        uint256 _newTotalAttoEthSold = _attoEthSold.add(_attoEthAmount);
+        uint256 _repPriceInAttoEth = (10**36) / _ethPriceInAttoRep;
+        currentLowerBoundRepPriceInAttoEth = currentLowerBoundRepPriceInAttoEth
+            .mul(_attoEthSold)
+            .add(_attoEthAmount.mul(_repPriceInAttoEth))
+            .div(_newTotalAttoEthSold);
+        currentAttoEthBalance = currentAttoEthBalance.sub(_attoEthAmount);
+        currentRepPrice = currentLowerBoundRepPriceInAttoEth.add(currentUpperBoundRepPriceInAttoEth) / 2;
+        return true;
+    }
+
+    function tradeEthForRep(uint256 _attoRepAmount) public initializeNewAuctionIfNeeded payable returns (bool) {
+        require(currentAttoRepBalance > 0);
+        require(_attoRepAmount > 0);
+        _attoRepAmount = _attoRepAmount.min(currentAttoRepBalance);
+        uint256 _repPriceInAttoEth = getRepSalePriceInAttoEth();
+        uint256 _attoEthCost = _attoRepAmount.mul(_repPriceInAttoEth) / 10**18;
+        // This will raise an exception if insufficient ETH was sent
+        msg.sender.transfer(msg.value.sub(_attoEthCost));
+        reputationToken.transfer(msg.sender, _attoRepAmount);
```

Comment:
I think the auction would be "better" if it was a uniform price auction instead (everyone gets the same price of the last bidder).  I think we should weigh the benefits of a uniform price auction (encourages people to bid early rather than playing a game of chicken) against the costs (code complexity).

As far as implementation, my initial thinking is that people trade ETH/REP for tokens that represent how much ETH/REP they contributed, then when the auction ends (due to amount of `ETH * current_ETH_price >= REP_available` or `REP * current_REP_price >= ETH_available` everyone gets a share of the ETH/REP available proportionate to their share of the auction tokens they hold.  I don't think code complexity will go up that much, though would need a prototype to verify that.

Reply:
Then we will need some sort of a threshold for when we consider the auction ended, and what is considered "dust" . Very commonly, systems are left with tiny amounts that no longer make sense to purchase. If i were using this system and wanted to buy a lot, I would buy up to what is considered dust + 1, forcing time to go lower before someone decides that it is worth it to buy that last batch

Reply:
In a uniform price multiunit auction the auction ends when the capital in escrow is greater than `lot * price`.  Since a price of `0` means the lot sells for `0`, this means the auction will _always_ fully clears the whole lot (exception for if _no one_ bids).